File: automedal/advisor/budget.py
"""Per-iteration token budget + junction gating for the advisor.

Process-global state: `reset_iteration_budget()` is called at the top of each
loop iteration; `consume_tokens(n)` is called inside `consult()` after each
successful advisor call. `remaining_tokens()` short-circuits further calls
once the cap is reached.
"""

from __future__ import annotations

import os

_STATE: dict[str, int] = {"used_this_iter": 0}


def _env_bool(key: str, default: bool) -> bool:
    raw = os.environ.get(key)
    if raw is None:
        return default
    return raw.strip().lower() in ("1", "true", "yes", "on")


def _env_int(key: str, default: int) -> int:
    try:
        return int(os.environ.get(key, default))
    except (TypeError, ValueError):
        return default


def reset_iteration_budget() -> None:
    _STATE["used_this_iter"] = 0


def consume_tokens(n: int) -> None:
    _STATE["used_this_iter"] = int(_STATE["used_this_iter"]) + max(0, int(n))


def remaining_tokens() -> int:
    cap = _env_int("AUTOMEDAL_ADVISOR_MAX_TOKENS_PER_ITER", 8000)
    return max(0, cap - int(_STATE["used_this_iter"]))


def budget_state() -> dict:
    cap = _env_int("AUTOMEDAL_ADVISOR_MAX_TOKENS_PER_ITER", 8000)
    used = int(_STATE["used_this_iter"])
    return {"used_this_iter": used, "cap_per_iter": cap, "remaining": max(0, cap - used)}


def _junctions_allowed() -> set[str]:
    raw = os.environ.get("AUTOMEDAL_ADVISOR_JUNCTIONS") or "stagnation,audit,tool"
    return {s.strip() for s in raw.split(",") if s.strip()}


def is_enabled(junction: str | None = None) -> bool:
    """Master flag + (optional) per-junction allowlist."""
    if not _env_bool("AUTOMEDAL_ADVISOR", False):
        return False
    if junction is None:
        return True
    return junction in _junctions_allowed()

File: internal/advisor/budget.go
// Per-iteration token budget + junction gating for the advisor.
//
// Process-global state: ResetIterationBudget() at the top of each loop
// iteration; ConsumeTokens(n) inside Consult after each successful call.
// RemainingTokens() short-circuits further calls once the cap is reached.
//
// Mirrors automedal/advisor/budget.py.
package advisor

import (
	"os"
	"strconv"
	"strings"
	"sync"
)

const (
	envEnabled       = "AUTOMEDAL_ADVISOR"
	envJunctions     = "AUTOMEDAL_ADVISOR_JUNCTIONS"
	envCapPerIter    = "AUTOMEDAL_ADVISOR_MAX_TOKENS_PER_ITER"
	envCapPerConsult = "AUTOMEDAL_ADVISOR_MAX_TOKENS_PER_CONSULT"
	envBaseURL       = "AUTOMEDAL_ADVISOR_BASE_URL"
	envModel         = "AUTOMEDAL_ADVISOR_MODEL"
	envAPIKey        = "OPENCODE_API_KEY"
)

const (
	defaultJunctions     = "stagnation,audit,tool"
	defaultCapPerIter    = 8000
	defaultCapPerConsult = 2000
	defaultBaseURL       = "https://opencode.ai/zen/go/v1"
	defaultModel         = "kimi-k2.6"
)

var (
	budgetMu      sync.Mutex
	usedThisIter  int
)

func envBool(key string, def bool) bool {
	raw := strings.TrimSpace(os.Getenv(key))
	if raw == "" {
		return def
	}
	switch strings.ToLower(raw) {
	case "1", "true", "yes", "on":
		return true
	}
	return false
}

func envInt(key string, def int) int {
	raw := os.Getenv(key)
	if raw == "" {
		return def
	}
	n, err := strconv.Atoi(raw)
	if err != nil {
		return def
	}
	return n
}

func envStr(key, def string) string {
	v := os.Getenv(key)
	if v == "" {
		return def
	}
	return v
}

// ResetIterationBudget zeros the per-iteration token tally.
func ResetIterationBudget() {
	budgetMu.Lock()
	defer budgetMu.Unlock()
	usedThisIter = 0
}

// ConsumeTokens adds n to the per-iteration tally (clamped ≥0).
func ConsumeTokens(n int) {
	if n < 0 {
		n = 0
	}
	budgetMu.Lock()
	defer budgetMu.Unlock()
	usedThisIter += n
}

// RemainingTokens returns max(0, cap - used).
func RemainingTokens() int {
	cap := envInt(envCapPerIter, defaultCapPerIter)
	budgetMu.Lock()
	defer budgetMu.Unlock()
	r := cap - usedThisIter
	if r < 0 {
		return 0
	}
	return r
}

// BudgetState reports (used, cap, remaining).
func BudgetState() map[string]int {
	cap := envInt(envCapPerIter, defaultCapPerIter)
	budgetMu.Lock()
	defer budgetMu.Unlock()
	rem := cap - usedThisIter
	if rem < 0 {
		rem = 0
	}
	return map[string]int{
		"used_this_iter": usedThisIter,
		"cap_per_iter":   cap,
		"remaining":      rem,
	}
}

func junctionsAllowed() map[string]bool {
	raw := envStr(envJunctions, defaultJunctions)
	out := map[string]bool{}
	for _, s := range strings.Split(raw, ",") {
		if s = strings.TrimSpace(s); s != "" {
			out[s] = true
		}
	}
	return out
}

// IsEnabled reports whether the advisor is on at all (and, if junction
// is non-empty, whether that junction is in the allowlist).
func IsEnabled(junction string) bool {
	if !envBool(envEnabled, false) {
		return false
	}
	if junction == "" {
		return true
	}
	return junctionsAllowed()[junction]
}
